### PR TITLE
C++: Abstractions for treating qualifiers as parameters in IR

### PR DESCRIPTION
**Pull request by MathiasVP**
Fixes https://github.com/github/codeql-c-analysis-team/issues/186.

In https://github.com/github/codeql/pull/3571 we made it possible to treat `this` as an argument, but there are still a couple of places where we need to distinguish between qualifiers and parameters (and between `this` and arguments). These places are:

- When accessing the arguments supplied to a `CallInstruction` (which has a `getPositionalArgument` and `getThisArgument` predicates)
- When working with instructions that initialize parameters (where initializers of positional parameters have a result for the `getParameter` predicate, but instructions initializing `this` need to refer to the `IRVariable`). The same problem also exists for `ReturnIndirectionInstruction`.

This PR adds a new predicate for each of these scenarios:
- `CallInstruction` now has a `getPositionalOrThisArgumentOperand` (and `getPositionalOrThisArgument`) predicates that, given an index, gives back the argument `operand` (and `instruction`) with the given index, or when the index is `-1`, gives back the operand used as the qualifier.
- `InitializeParameterInstruction` and `InitializeIndirectionInstruction ` now have an `isParameterOrQualifierIndex` predicate that holds if the instruction initializes a positional parameter with a given index, or initialize `this` with the index `-1`.
  Similarly, `ReturnIndirectionInstruction` has a new predicate `isParameterOrThisIndirection` with similar semantics.

### CODE REVIEW

**Comment by MathiasVP** [comment]
CPP-difference started: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1563/

**Comment by rdmarsh2** [comment]
I think we want to guide users towards using these by default. Could `getPositionalOrThisArgumentOperand` just be `getArgumentOperand`? Similarly, would `isParameterOrQualifierIndex` be better as `hasIndex`?

**Comment by MathiasVP** [comment]
> I think we want to guide users towards using these predicates by default, which usually means giving them names that are simpler than their alternatives. Could `getPositionalOrThisArgumentOperand` just be `getArgumentOperand`? Similarly, would `isParameterOrQualifierIndex` be better as `hasIndex`?

I think those are some excellent names! I've changed the names to match your suggestions now. I had to get rid of an unused predicate on DataFlowCall in `DataFlowPrivate` since the name was clashing with your suggestion (and I think removing that predicate is okay since I don't think we expect users to use `DataFlowCall`?)

**Comment by MathiasVP** [comment]
> CPP-difference started: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1563/

CPP-difference looks good (for some reason there's a 10% reduction in build/extraction/import time on openjdk?)

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll [hunk 1]
```diff
@@ -775,6 +796,16 @@ class ReturnIndirectionInstruction extends VariableInstruction {
...
+
+  /**
+   * Holds if this instruction is the return indirection for the parameter with index `index`, or
+   * if this instruction is the return indirection for `this` and `index` is `-1`.
+   */
+  final predicate hasIndex(int index) {
+    index >= 0 and index = this.getParameter().getIndex()
+    or
+    index = -1 and this.isThisIndirection()
+  }
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll:808]: The implementation of this predicate looks a lot like `getArgumentOperand`, but only the latter has `pragma[noinline]` on it. Is there a good reason to put `pragma[noinline]` on one but not the other?

**Review by jbj — APPROVED** [review]
Looks great! I'm happy with the names too.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll:808]: Hm, no. I suspect the `pragma[noinline]` is an artifact of me copy/pasting from `getPositionalArgumentOperand`. I'll just double-check that I can safely remove the new `pragma[noinline]`s.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll:808]: Hm... looking at [this comment you made](https://github.com/github/codeql/pull/4637#discussion_r519897265), I'd actually feel more comfortable adding a `pragma[noinline]` to the `hasIndex ` predicates. It doesn't seem to be necessary right now, but if it avoids a future bad join-order we might as well put it there now. Would that be reasonable?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll:811]: Yes.

**Review by jbj — APPROVED** [review]

**Pull request merged**